server/events/runtime/init_step_runner.go

```diff
@@ -16,14 +20,55 @@ func (i *InitStepRunner) Run(ctx models.ProjectCommandContext, extraArgs []strin
 	if ctx.TerraformVersion != nil {
 		tfVersion = ctx.TerraformVersion
 	}
-	terraformInitCmd := append([]string{"init", "-input=false", "-no-color", "-upgrade"}, extraArgs...)
+
+	terraformInitVerb := []string{"init"}
+	terraformInitArgs := []string{"-input=false"}
 
 	// If we're running < 0.9 we have to use `terraform get` instead of `init`.
 	if MustConstraint("< 0.9.0").Check(tfVersion) {
 		ctx.Log.Info("running terraform version %s so will use `get` instead of `init`", tfVersion)
-		terraformInitCmd = append([]string{"get", "-no-color", "-upgrade"}, extraArgs...)
+		terraformInitVerb = []string{"get"}
+		terraformInitArgs = []string{}
+	}
+
+	terraformInitArgs = append(terraformInitArgs, "-no-color")
+
+	lockfilePath := filepath.Join(path, ".terraform.lock.hcl")
+	if MustConstraint("< 0.14.0").Check(tfVersion) || fileDoesNotExists(lockfilePath) {
+		terraformInitArgs = append(terraformInitArgs, "-upgrade")
+	}
+
+	// work if any of the core args have been overridden
+	finalArgs := []string{}
+	usedExtraArgs := []string{}
+	for _, arg := range terraformInitArgs {
+		override := ""
+		prefix := arg
+		argSplit := strings.Split(arg, "=")
+		if len(argSplit) == 2 {
+			prefix = argSplit[0]
+		}
+		for _, extraArg := range extraArgs {
+			if strings.HasPrefix(extraArg, prefix) {
+				override = extraArg
+			}
+		}
+		if override != "" {
+			finalArgs = append(finalArgs, override)
+			usedExtraArgs = append(usedExtraArgs, override)
+		} else {
+			finalArgs = append(finalArgs, arg)
+		}
+	}
+	// add any extra args that are not overrides
+	for _, extraArg := range extraArgs {
+		if !stringInSlice(usedExtraArgs, extraArg) {
+			finalArgs = append(finalArgs, extraArg)
+		}
 	}
 
```

Comment:
Mind moving this to another function? Would be useful to use for the conftest extra args implementation as well also this function will be a bit cleaner that way.